src/EditorFeatures/Core/StackTraceExplorer/StackTraceExplorerUtilities.cs

```diff
@@ -0,0 +1,211 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.FindUsages;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.StackFrame;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.StackTraceExplorer;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.StackTraceExplorer
+{
+    internal static class StackTraceExplorerUtilities
+    {
+        public static async Task<DefinitionItem?> GetDefinitionAsync(Solution solution, StackFrameCompilationUnit compilationUnit, StackFrameSymbolPart symbolPart, CancellationToken cancellationToken)
+        {
+            // MemberAccessExpression is [Expression].[Identifier], and Identifier is the 
+            // method name.
+            var typeExpression = compilationUnit.MethodDeclaration.MemberAccessExpression.Left;
+            var fullyQualifiedTypeName = typeExpression.ToString();
+            var typeName = typeExpression is StackFrameQualifiedNameNode qualifiedName
+                ? qualifiedName.Right.ToString()
+                : typeExpression.ToString();
+
+            RoslynDebug.AssertNotNull(fullyQualifiedTypeName);
+
+            var methodIdentifier = compilationUnit.MethodDeclaration.MemberAccessExpression.Right;
+            var methodTypeArguments = compilationUnit.MethodDeclaration.TypeArguments;
+            var methodArguments = compilationUnit.MethodDeclaration.ArgumentList;
+
+            var methodName = methodIdentifier.ToString();
+
+            //
+            // Do a first pass to find projects with the type name to check first 
+            //
+            using var _ = PooledObjects.ArrayBuilder<Project>.GetInstance(out var candidateProjects);
+            foreach (var project in solution.Projects)
+            {
+                if (!project.SupportsCompilation)
+                {
+                    continue;
+                }
+
+                var containsSymbol = await project.ContainsSymbolsWithNameAsync(
+                    (name) => name == typeName,
+                    SymbolFilter.Type,
+                    cancellationToken).ConfigureAwait(false);
+
+                if (containsSymbol)
+                {
+                    var matchingMethods = await GetMatchingMembersFromCompilationAsync(project).ConfigureAwait(false);
+                    if (matchingMethods.Any())
+                    {
+                        return await GetDefinitionAsync(matchingMethods[0]).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    candidateProjects.Add(project);
+                }
+            }
+
+            //
+            // Do a second pass to check the remaining compilations
+            // for the symbol, which may be a metadata symbol in the compilation
+            //
+            foreach (var project in candidateProjects)
```

Comment:
the candidateProjects local seems unnecessary.

Reply:
This just keeps us from getting the compilation twice and searching on a project that we already searched. Could do without... 🤷 the compilation would already be computed so theoretically the expensive part is done.